pkg/formula/runner/local/pre_run_build.go

```
@@ -0,0 +1,202 @@
+
+		// Remove /bin dir to force formula rebuild in next execution
+		if err := this.dir.Remove(binFilePath); err != nil {
+			return err
+		}
+
+		return err
+	}
+
+	s.Success(prompt.Green(messageBuildSuccess))
+	return nil
+}
+
+func (this PreRunBuilder) mustBuild(formulaPath string, binFilePath string) bool {
+	hasExecutable := this.file.Exists(binFilePath)
+	if !hasExecutable {
+		return true
+	}
+
+	// Test: should not check for changes when Bin doesn't exist
+	// Test: should not be blocked by errors
+	// Test: should proceed execution when no changes are made
+	// Test: should proceed execution when user chooses not to rebuild
+	// Test: should rebuild when requested by user
+	hasFormulaChanged, err := this.hasFormulaChanged(formulaPath)
+	if err != nil {
+		fmt.Println(prompt.Yellow(messageChangeError))
+		return false
+	}
+	if !hasFormulaChanged {
+		return false
+	}
+
+	ans, err := prompt.NewSurveyBool().Bool(messageChangePrompt, []string{messageYes, messageNo})
+	if err != nil {
+		return false
+	}
+	return ans
+}
+
+func (this PreRunBuilder) hasFormulaChanged(formulaPath string) (bool, error) {
+	// Test: should always run updateHash, even when getPreviousHash fails
+	previousHash, getErr := this.getPreviousHash(formulaPath)
+	currentHash, setErr := this.updateHash(formulaPath)
+
+	if getErr != nil {
+		return false, nil
+	}
+	if setErr != nil {
+		return false, setErr
+	}
+
+	return previousHash != currentHash, nil
+}
+
+func (this PreRunBuilder) getPreviousHash(formulaPath string) (string, error) {
+	filePath := this.getHashPath(formulaPath)
+
+	hash, err := ioutil.ReadFile(filePath)
+	if err != nil {
+		return "", err
+	}
+
+	return string(hash), nil
+}
+
+func (this PreRunBuilder) updateHash(formulaPath string) (hash string, err error) {
+	filePath := this.getHashPath(formulaPath)
+	currentHash, err := dirhash.HashDir(formulaPath, "", dirhash.DefaultHash)
+	if err != nil {
+		return "", err
+	}
+
+	_ = os.Mkdir(this.getHashDir(), os.ModePerm)
```

```stream.DirCreater``` can be a useful here.
why ignore the error?

---

If it fails because the directory already exists, we still want to continue creating the file

If it fails due to something else, the directory will not be created and the Write below it will fail (and that error is treated)

Is that ok or do you suggest another approach?

---

no problem to me